File: scrapeless-sdks/node/lib/bridge.js
'use strict';

/**
 * The JavaScript half: a buffering listener installed before page scripts run.
 *
 * Playwright cannot receive the extension's CustomEvents itself — they are
 * dispatched at the page's window. So this listener is installed into every
 * document at document_start via addInitScript, it records what it hears, and
 * Node reads it back with evaluate().
 *
 * Installing at document_start is the whole trick. The extension dispatches
 * scrapeless:onDetection almost immediately on a cache hit, so a listener added
 * after page.goto() returns would routinely hear nothing at all — and a silent
 * listener is indistinguishable from a clean page.
 *
 * This text is byte-identical to scrapeless-sdks/bridge.js in every SDK; a test
 * asserts it, because a drifted copy would fail silently.
 */

const INIT_SCRIPT = "(() => {\n  if (window.__scrapelessSdkBridge) { return; }\n  const EVENTS = ['ready', 'onStart', 'onProgress', 'onHooksComplete',\n                  'onWindowPropsComplete', 'onDetection', 'onError'];\n  const last = Object.create(null);\n  let waiters = [];\n\n  for (const name of EVENTS) {\n    window.addEventListener('scrapeless:' + name, (signal) => {\n      last[name] = (signal && signal.detail) || {};\n      if (name === 'onDetection') {\n        const pending = waiters;\n        waiters = [];\n        for (const settle of pending) { settle(last[name]); }\n      }\n    });\n  }\n\n  window.__scrapelessSdkBridge = {\n    snapshot: () => last.onDetection || null,\n    readyDetail: () => last.ready || null,\n    seen: () => Object.keys(last),\n    // Bounded: the page API ships disabled and the extension may be absent,\n    // and neither announces itself. Resolve null rather than hang forever.\n    awaitDetection: (timeoutMs) => new Promise((resolve) => {\n      if (last.onDetection) { resolve(last.onDetection); return; }\n      let settled = false;\n      const finish = (value) => {\n        if (settled) { return; }\n        settled = true;\n        resolve(value || null);\n      };\n      waiters.push(finish);\n      if (timeoutMs > 0) { setTimeout(() => finish(null), timeoutMs); }\n    })\n  };\n})();\n";

const AWAIT_DETECTION = '(ms) => window.__scrapelessSdkBridge.awaitDetection(ms)';
const SNAPSHOT = '() => window.__scrapelessSdkBridge && window.__scrapelessSdkBridge.snapshot()';
const READY = '() => window.__scrapelessSdkBridge && window.__scrapelessSdkBridge.readyDetail()';
const INSTALLED = '() => Boolean(window.__scrapelessSdkBridge)';

/**
 * Flags Chrome needs to actually load an unpacked extension. Both are required:
 * --load-extension is ignored unless the extension is also exempted from the
 * disable list.
 */
function chromiumExtensionArgs(extensionPath) {
    return [
        '--disable-extensions-except=' + extensionPath,
        '--load-extension=' + extensionPath
    ];
}

module.exports = {
    INIT_SCRIPT, AWAIT_DETECTION, SNAPSHOT, READY, INSTALLED, chromiumExtensionArgs
};

File: scrapeless-sdks/rust/src/bridge.rs
//! The JavaScript half: a buffering listener installed before page scripts run.
//!
//! Installing at document_start is the whole trick. The extension dispatches
//! `scrapeless:onDetection` almost immediately on a cache hit, so a listener
//! added after navigation would routinely hear nothing — and a silent listener
//! is indistinguishable from a clean page.
//!
//! Byte-identical to `scrapeless-sdks/bridge.js` in every SDK; the Node suite
//! asserts it, because a drifted copy fails silently otherwise.

/// Installed via `Page.addScriptToEvaluateOnNewDocument`.
pub const INIT_SCRIPT: &str = r#"
(() => {
  if (window.__scrapelessSdkBridge) { return; }
  const EVENTS = ['ready', 'onStart', 'onProgress', 'onHooksComplete',
                  'onWindowPropsComplete', 'onDetection', 'onError'];
  const last = Object.create(null);
  let waiters = [];

  for (const name of EVENTS) {
    window.addEventListener('scrapeless:' + name, (signal) => {
      last[name] = (signal && signal.detail) || {};
      if (name === 'onDetection') {
        const pending = waiters;
        waiters = [];
        for (const settle of pending) { settle(last[name]); }
      }
    });
  }

  window.__scrapelessSdkBridge = {
    snapshot: () => last.onDetection || null,
    readyDetail: () => last.ready || null,
    seen: () => Object.keys(last),
    // Bounded: the page API ships disabled and the extension may be absent,
    // and neither announces itself. Resolve null rather than hang forever.
    awaitDetection: (timeoutMs) => new Promise((resolve) => {
      if (last.onDetection) { resolve(last.onDetection); return; }
      let settled = false;
      const finish = (value) => {
        if (settled) { return; }
        settled = true;
        resolve(value || null);
      };
      waiters.push(finish);
      if (timeoutMs > 0) { setTimeout(() => finish(null), timeoutMs); }
    })
  };
})();
"#;

pub const SNAPSHOT: &str =
    "(window.__scrapelessSdkBridge ? window.__scrapelessSdkBridge.snapshot() : null)";
pub const READY: &str =
    "(window.__scrapelessSdkBridge ? window.__scrapelessSdkBridge.readyDetail() : null)";
pub const INSTALLED: &str = "Boolean(window.__scrapelessSdkBridge)";

/// CDP evaluates an expression, not a function, so the wait is built as an IIFE
/// with the budget interpolated.
pub fn await_expression(timeout_ms: u64) -> String {
    format!(
        "(window.__scrapelessSdkBridge ? window.__scrapelessSdkBridge.awaitDetection({}) : null)",
        timeout_ms
    )
}

/// Flags Chrome needs to actually load an unpacked extension. Both are required:
/// `--load-extension` is ignored unless the extension is also exempted from the
/// disable list.
pub fn chromium_extension_args(extension_path: &str) -> Vec<String> {
    vec![
        format!("--disable-extensions-except={}", extension_path),
        format!("--load-extension={}", extension_path),
    ]
}
